coq_kernel: init at 1.6.0

thomasjm opened the pull request:
###### Description of changes

This is a Jupyter kernel for Coq, the formal proof management system.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

thomasjm committed:
coq_kernel: init at 1.6.0

thomasjm commented:
@SuperSandro2000 any chance I could trouble you to review this? I think it should be straightforward, I did it the same way as other Jupyter kernels I've packaged in the past.

SuperSandro2000 commented on pkgs/applications/editors/jupyter-kernels/coq/kernel.nix:
```diff
@@ -0,0 +1,31 @@
+    description = "Jupyter kernel for Coq";
+    license = licenses.asl20;
+    maintainers = with maintainers; [ thomasjm ];
+    platforms = platforms.all;
```
```suggestion
```
the default platform of buildPythonPackage is similar to unix which is smaller then all.

thomasjm replied:
Done

SuperSandro2000 commented on pkgs/applications/editors/jupyter-kernels/coq/kernel.nix:
```diff
@@ -0,0 +1,31 @@
+    sha256 = "sha256-+Pp51cxeqjg5MW4CEccNWVjNcY9iyFNATIEage9RWJ0=";
+  };
+
+  propagatedBuildInputs = (with python3.pkgs; [ipykernel future]) ++ [coq];
```
```suggestion
  propagatedBuildInputs = (with python3.pkgs; [ ipykernel future ]) ++ [ coq ];
```

thomasjm replied:
Done

SuperSandro2000 commented on pkgs/applications/editors/jupyter-kernels/coq/kernel.nix:
```diff
@@ -0,0 +1,31 @@
+}:
+
+python3.pkgs.buildPythonPackage rec {
+  pname = "coq_jupyter";
```
```suggestion
  pname = "coq-jupyter";
```

thomasjm replied:
Done

SuperSandro2000 reviewed (commented):
@siraben don't you know coq?

siraben commented:
> @siraben don't you know coq?

I do, but I'm not very familiar with Jupyter. I'll review this when I have some time.

thomasjm commented:
Any chance we can get this merged soon? Coq familiarity isn't really required, you just have to start up the Jupyter notebook and confirm that it successfully runs Coq code. It works for me when doing the following:

```console
nix run nixpkgs#nixpkgs-review pr 171202
$(nix-build -E 'with import ./. {}; jupyter.override { definitions = { coq = coq-kernel.definition; }; }')/bin/jupyter-notebook
```

thomasjm commented:
@SuperSandro2000 @siraben ping again on this.

thomasjm commented:
I was actually looking at the metakernel breakage yesterday and realized `coq_jupyter` doesn't actually require it. Just removed it and rebased on master.

thomasjm commented:
> How would one go about adding coqPackages?

I'm not sure, the manual seems to only describe how to add new ones. If you figure it out please lmk!

thomasjm commented:
Actually, it seems like `coqPackages` are meant to be installed alongside `coq` in the same environment. For example, if I build `coqPackages.ceres`, it produces a lib path `result/lib/coq/8.16/user-contrib/Ceres`. Presumably `coq` will search these kinds of paths when you do an `Import`.

However, it's not working too well for me. I tried installing `coq` alongside `coqPackages.ceres`, and the import still failed. I notice `coq` alone produces `lib/coq/user-contrib` so perhaps there's a slight mismatch here.

thomasjm commented:
Aha I figured it out, based on the bullet point about `extraInstallFlags` here: https://nixos.org/manual/nixpkgs/stable/#coq-packages-attribute-sets-coqpackages.

Just pushed a commit adding a `definitionWithPackages` function.

skeuchel commented:
> Just pushed a commit adding a definitionWithPackages function.

That should be pretty useful. Pinning version is common in this space, so could you also make the coq version configurable? I would like to pass say `coq_8_15` along with some `coqPackages_8_15`. I will test it again after than and then it should be ready before more feature creep sets in ;)

thomasjm commented:
I think it's already configurable via `override`, which AFAIK is the standard pattern.

thomasjm commented:
What I'd really like to do is open a PR to either attach the `coqPackages` for a given version to the `coq`, or even better, add a `coq.withPackages` function.

skeuchel commented:
I've tested it with adding some `coqPackages` and also overriding the coq version. Everything works so it LGTM.

thomasjm commented:
Great! Let's see if we can get it merged now.

SuperSandro2000 commented on pkgs/applications/editors/jupyter-kernels/coq/default.nix:
```diff
@@ -0,0 +1,67 @@
+      --suffix PATH : $coq/bin
+  '';
+
+  sizedLogo = size: stdenv.mkDerivation {
+    pname = "coq-logo-${size}x${size}.png";
+    inherit (coq) version;
+
+    src = coq.src;
+
+    buildInputs = [ imagemagick ];
+
+    dontConfigure = true;
+    dontInstall = true;
+
+    buildPhase = ''
+      convert ./ide/coqide/coq.png -resize ${size}x${size} $out
+    '';
```
Please generate all sizes in one derivation. Also this should use runCommandNoCC

thomasjm replied:
Huh? `runCommandNoCC` isn't used all in the current master branch of Nixpkgs. It's aliased to `runCommand`. I did it but I think I'd better change it back if you don't mind.

All sizes are in one derivation now.

thomasjm replied:
I've changed it back in my latest commit, lmk if you want to discuss further.

SuperSandro2000 commented on pkgs/applications/editors/jupyter-kernels/coq/default.nix:
```diff
@@ -0,0 +1,67 @@
+  } ''
+    mkdir -p $out/bin
+
+    makeWrapper $python/bin/python $out/bin/coq-kernel \
```
```suggestion
    makeWrapper ${python.interpreter} $out/bin/coq-kernel \
```

SuperSandro2000 commented on pkgs/applications/editors/jupyter-kernels/coq/default.nix:
```diff
@@ -0,0 +1,67 @@
+  launcher = runCommand "coq-kernel-launcher" {
+    inherit coq;
+    python = python3.withPackages (ps: [ ps.traitlets ps.jupyter_core ps.ipykernel kernel ]);
+    buildInputs = [ makeWrapper ];
```
makeWrapper goes to nativeBuildInputs

thomasjm replied:
Done

SuperSandro2000 commented on pkgs/applications/editors/jupyter-kernels/coq/default.nix:
```diff
@@ -0,0 +1,67 @@
+
+rec {
+  launcher = runCommand "coq-kernel-launcher" {
+    inherit coq;
```
```suggestion
```

thomasjm replied:
I swear I was told the opposite once in a previous review :)

SuperSandro2000 commented on pkgs/applications/editors/jupyter-kernels/coq/default.nix:
```diff
@@ -0,0 +1,67 @@
+
+    makeWrapper $python/bin/python $out/bin/coq-kernel \
+      --add-flags "-m coq_jupyter" \
+      --suffix PATH : $coq/bin
```
```suggestion
      --suffix PATH : ${coq}/bin
```

thomasjm commented:
@SuperSandro2000 ? This is still ready to go

thomasjm commented:
Hi @jonringer, I'm hoping to find a reviewer for this and a few other Jupyter-related PRs. I saw you committed the initial IRuby PR so I was wondering if you could take a look here?

thomasjm commented:
I've just updated this to be consistent with the latest Jupyter stuff. I think it's ready to merge!

GTrunSec commented:
Could you rebase your changes with the latest nixpkgs?

thomasjm commented:
Done @GTrunSec

GTrunSec commented:
Test: `nix run --impure --expr 'with import ./. {}; jupyter-console.withSingleKernel (coq-kernel.definitionWithPackages [coqPackages.bignums])'`

```
nix run --impure --expr 'with import ./. {}; jupyter-console.withSingleKernel (coq-kernel.definitionWithPackages [coqPackages.bignums])'
Jupyter console 6.6.3

The Coq Proof Assistant, version 8.18.0
compiled with OCaml 4.14.1

In [1]: Compute (1 + 1).
Out[2]:
     = 2
     : nat
```
Thanks, LGTM

thomasjm commented:
Cool! I guess now we ping @teto that this kernel is ready to go?

GTrunSec commented:
we should ping @natsukium jupyter team.

teto commented on pkgs/applications/editors/jupyter-kernels/coq/default.nix:
```diff
@@ -0,0 +1,58 @@
+, coq
+, imagemagick
+, python3
+}:
```
could you move the file to pkgs/by-name as explained here https://github.com/NixOS/nixpkgs/blob/master/pkgs/by-name/README.md ? there are plenty of examples in the latest PRs

thomasjm replied:
Hmm, I think that may not be the best idea here, for the following two reasons:

1. I think there are significant advantages to keeping the Jupyter kernels together in a common folder. In the past they've been somewhat fragmented through the codebase, and it makes it hard to refactor them together in a consistent way. I'd like to do more PRs to improve their cleanliness and ergonomics, and it may involve factoring out common code.
2. This `pkgs/by-name` effort sounds like it is for top-level packages. I wouldn't describe these kernels as top-level packages. I think we want to work toward a world where the kernels are consolidated into a common little package set and there's some nice machinery for launching notebooks/console/etc. with your desired subset of kernels. My `jupyter-console.withSingleKernel` PR was a first step in this direction. Perhaps the full set of kernels should be exposed in a top-level `jupyter-kernels` attrset, but they're not much use on their own at the top level.

teto replied:
Fair enough. I kinda like to have them all in a subfolder too.

teto commented on pkgs/applications/editors/jupyter-kernels/coq/default.nix:
```diff
@@ -0,0 +1,58 @@
+
+# Jupyter console:
+# nix run --impure --expr 'with import <nixpkgs> {}; jupyter-console.withSingleKernel coq-kernel.definition'
+
```
you put this in every one of your contribution but I am not a fan as it can get outdated pretty quickly. I think it's best to add a test in passthru.test that just tries to build the package as is done for neovim (if you want an example) and/or improve the documentation: either in the manual or in a README for jupyter that you could link here as a comment.
The `with import <nixpkgs> {};` tends to be outdated as a syntax so better not spread it.

thomasjm replied:
Tests already exist; the kernels are built as part of the `jupyter-all` derivation.

I agree this isn't the best place for documentation. As part of the effort to make the kernels more organized and homogeneous, I'd like to have nice ways of launching them which are explained in the manual. As things are still a little fragmentary, I think it's useful to keep these comments at the top explaining how to launch each particular kernel. But removing them is definitely part of my multi-step plan :)

Would you be happier with `with import ./. {};` ?

thomasjm replied:
Thanks!
